### PR TITLE
always process repair pong messages

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem
Pongs are required for further repair processing and could be dropped by rate limiting.

#### Summary of Changes
Always process repair pong messages.

Fixes #

### CODE REVIEW

File: core/src/serve_repair.rs [hunk 1]
```diff
@@ -1183,6 +1214,68 @@ mod tests {
...
+        assert!(ServeRepair::is_pong(&pkt));
+
+        let ping = Ping::new_rand(&mut rng, &keypair).unwrap();
+        let repair_ping = RepairResponse::Ping(ping);
+        let pkt = Packet::from_data(None, repair_ping).unwrap();
+        assert!(!ServeRepair::is_pong(&pkt));
+    }
+
+    #[test]
+    fn test_retain_pongs() {
```
Comment (HaoranYi) [inline core/src/serve_repair.rs:1235]: +1 for the test

**Review by HaoranYi — APPROVED** [review]
lgtm. thanks!

**Comment by behzadnouri** [comment]
Once the initial ping/pong is established, then ping/pong cache proactively tries to extend that well before the token expires:
https://github.com/solana-labs/solana/blob/abfaf06e8/gossip/src/ping_pong.rs#L245-L247

And it makes several attempts to do this, and so each of these attempt failing is not significant because there will be more pings sent before expiration of token.

So does not seem to me dropping pongs should be an issue. do we actually have any evidence that this is an issue?

Worst case is that the requester is pinged again once the delay is over. If the server is already overwhelmed that it is dropping the requests, then the repair requests are not responded anyway.

**Comment by jbiseda** [comment]
> Once the initial ping/pong is established, then ping/pong cache proactively tries to extend that well before the token expires: https://github.com/solana-labs/solana/blob/abfaf06e8/gossip/src/ping_pong.rs#L245-L247
> 
> And it makes several attempts to do this, and so each of these attempt failing is not significant because there will be more pings sent before expiration of token.
> 
> So does not seem to me dropping pongs should be an issue. do we actually have any evidence that this is an issue?
> 
See #27958 to collect this data (for gossip) which would be used to collect this for repairs as well.  If nodes are bandwidth limited and dropping pongs when ping/pong checking is enabled this could have cause a substantial impact on repair processing until the pongs are included for processing.

> Worst case is that the requester is pinged again once the delay is over. If the server is already overwhelmed that it is dropping the requests, then the repair requests are not responded anyway.
> 
Dropping a pong packet is more impactful than other packets as all other repairs will be dropped until a pong is successfully processed. A dropped pong effectively stalls all repairs from that node for the ping rate limit duration.

> The cost of this approach is that allowing pongs to bypass rate limiting can easily be abused
I'm not sure this approach is worse than existing behavior. If this is a problem then we could also enforce some upper limit on the number of pongs which bypass the load shedding filter.

**Comment by jbiseda** [comment]
From the handful of validators on testnet reporting the new load shed/bandwidth drop metrics (these are max values):
Screen Shot 2022-09-28 at 2 24 47 PM

`5D1fNX` is the highest staked validator on testnet and is dropping a substantial number of packets during load shed. Seems like this could impede pong processing.

**Comment by behzadnouri** [comment]
> Dropping a pong packet is more impactful than other packets as all other repairs will be dropped until a pong is successfully processed. A dropped pong effectively stalls all repairs from that node for the ping rate limit duration.

I understand the goal. The question is that if the way ping/pong cache proactively extends tokens' lifetime before their `ttl` expiry here:
https://github.com/solana-labs/solana/blob/7469ce7ba/gossip/src/ping_pong.rs#L245-L247
is already enough to avoid that problem or not?

We can figure that out by looking at:
https://github.com/solana-labs/solana/blob/7469ce7ba/gossip/src/cluster_info.rs#L1909-L1911
in gossip and then an equivalent such metric in repair.

If it is shown that it is not sufficient, It might make sense to use different `ttl` and extension parameters for repair compared to gossip, similar to how you changed the `rate_limit_delay`.

We can also relax the rate limiting here a bit more:
https://github.com/solana-labs/solana/blob/7469ce7ba/core/src/serve_repair.rs#L459-L474
and allow more packets to reach here:
https://github.com/solana-labs/solana/blob/7469ce7ba/core/src/serve_repair.rs#L712-L719
but instead of breaking out of the loop once bandwidth is expired, process any remaining pong packets.

**Pull request closed without merging**